Gems/LmbrCentral/Code/Tests/AxisAlignedBoxShapeTest.cpp

```diff
@@ -164,4 +165,31 @@ namespace UnitTest
         EXPECT_TRUE(rayHit);
         EXPECT_NEAR(distance, 4.0f, 1e-2f);
     }
+
+    TEST_F(AxisAlignedBoxShapeTest, ShapeHasThreadsafeGetSetCalls)
+    {
+        // Verify that setting values from one thread and querying values from multiple other threads in parallel produces
+        // correct, consistent results.
+
+        // Create our axis-aligned box centered at 0 with our height and starting XY dimensions.
+        AZ::Entity entity;
+        CreateAxisAlignedBox(
+            AZ::Transform::CreateTranslation(AZ::Vector3::CreateZero()),
+            AZ::Vector3(ShapeThreadsafeTest::MinDimension, ShapeThreadsafeTest::MinDimension, ShapeThreadsafeTest::ShapeHeight), entity);
+
+        // Define the function for setting unimportant dimensions on the shape while queries take place.
+        auto setDimensionFn = [](AZ::EntityId shapeEntityId, float minDimension, uint32_t dimensionVariance, float height)
+        {
+            float x = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+            float y = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+
+            LmbrCentral::BoxShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::BoxShapeComponentRequestsBus::Events::SetBoxDimensions, AZ::Vector3(x, y, height));
+        };
+
+        // Run the test, which will run multiple queries in parallel with each other and with the dimension-setting function.
+        const int numIterations = 100000;
```

Comment:
Do we really need a 100000 iterations?
I wonder about the trade-off between trying to detect a race condition vs test time
Any thoughts @Kadino ?

Reply:
Agreed that this seems like an intense amount of executions. This type of race-condition fuzz-testing thrash is a bit of a paranoid way to test, so I would suggest limiting it to at most 2x the minimal set of iterations that load up the parallel threads and create the expected scenario. Better yet is to add specific tests for edge cases when they are found.

How long does the test currently take? I'd recommend targeting something lower than a full second, given how many cycles should execute. Given that it will get run in Jenkins many dozens of times per day, I expect a handful of iterations should be more than enough.

Adding a CLI parameter might reduce investigation time by a few minutes of compilation. Complicates the code a bit for a "maybe useful someday" scenario. I don't find that too compelling.

Reply:
**Short version:** 
Each added test takes < 500 ms right now. If this needs to be reduced further, let me know.

**Long version:**
"Do we really need a 100000 iterations? I wonder about the trade-off between trying to detect a race condition vs test time."
I picked 100,000 as a balance between as many iterations as possible and keeping the tests well under 1 second each. Here's how much time the tests take in a profile build on my PC:
```
QuadShapeTest.ShapeHasThreadsafeGetSetCalls (119 ms)
TubeShapeTest.ShapeHasThreadsafeGetSetCalls (88 ms)
DiskShapeTest.ShapeHasThreadsafeGetSetCalls (118 ms)
ReferenceComponentTests.ShapeHasThreadsafeGetSetCalls (226 ms)
AxisAlignedBoxShapeTest.ShapeHasThreadsafeGetSetCalls (116 ms)
BoxShapeTest.ShapeHasThreadsafeGetSetCalls (86 ms)
SphereShapeTest.ShapeHasThreadsafeGetSetCalls (96 ms)
CylinderShapeTest.ShapeHasThreadsafeGetSetCalls (95 ms)
CapsuleShapeTest.ShapeHasThreadsafeGetSetCalls (111 ms)
PolygonPrismShapeTest.ShapeHasThreadsafeGetSetCalls (488 ms)
```

Just for completeness, here are the numbers from a debug build. The only egregious one (IMO) is the PolygonPrism, which could probably benefit from some optimization in the DistanceFromPointSquared method:
```
QuadShapeTest.ShapeHasThreadsafeGetSetCalls (174 ms)
TubeShapeTest.ShapeHasThreadsafeGetSetCalls (427 ms)
DiskShapeTest.ShapeHasThreadsafeGetSetCalls (160 ms)
ReferenceComponentTests.ShapeHasThreadsafeGetSetCalls (776 ms)
AxisAlignedBoxShapeTest.ShapeHasThreadsafeGetSetCalls (209 ms)
BoxShapeTest.ShapeHasThreadsafeGetSetCalls (238 ms)
SphereShapeTest.ShapeHasThreadsafeGetSetCalls (110 ms)
CylinderShapeTest.ShapeHasThreadsafeGetSetCalls (234 ms)
CapsuleShapeTest.ShapeHasThreadsafeGetSetCalls (192 ms)
PolygonPrismShapeTest.ShapeHasThreadsafeGetSetCalls (2594 ms)
```

I'm happy to drop the iteration value further if this seems overly burdensome. If so, please recommend the time target to aim for.

"This type of race-condition fuzz-testing thrash is a bit of a paranoid way to test, so I would suggest limiting it to at most 2x the minimal set of iterations that load up the parallel threads and create the expected scenario. Better yet is to add specific tests for edge cases when they are found."

To be clear, this is my absolutely _least_ favorite way to test for race conditions. I _much_ prefer setting up lockstep-style tests. However, for the shape code, I couldn't find a better way to validate the existence and removal of the race conditions that I could clearly see existing through manual code inspection. Specifically, what this is detecting is the brokenness that can occur when (for example) `CapsuleShape::CapsuleIntersectionDataCache::UpdateIntersectionParamsImpl` is updating the 6 cached values it has while `CapsuleShape::SetRadius` is changing the configured radius in-between different calculations in the update that use the radius. The race condition failure occurs when the updated cached values weren't all calculated with the same configuration values, which produces a result that doesn't match either the before _or_ the after state of the capsule shape change. This condition only occurs in the capsule case when the radius changes in-between those few lines of code.

This test may seem paranoid, but it successfully detected the race condition 100% of the time on 4/10 of the shapes prior to the fixes, and 0% of the time after the fixes. No other automated tests were detecting the problem before this was written, and this is a demonstrable problem, so I believe it to be a valid and useful test for the specific bug it is detecting. I did attempt to tune it to minimize the overall runtime while maximizing the chance of successful detection. My goal was to tune this to 100% detection as best as I could, because _intermittent_ failures on AR runs due to thrash-based unit tests is never enjoyable.

It's unfortunately also worth noting that this unit test did _not_ successfully detect the race condition on 6/10 of the shapes even prior to my fixes, mostly because their logic and number of cached values are simple enough that the changes are staying _effectively_ atomic, even if the code has no mechanisms to enforce that. I still left those tests in as well though, because they at least establish a baseline to ensure that changes to the shape code don't start causing this condition to fail.

Reply:
I'm interested in the threshold where the detected cases stop being reliably detected (on say 1/10 runs) on your local environment. If that currently occurs around 50k iterations, then this is a reasonable test.

Part of me would love to be a purist and say "no, never fuzz!" but the more important testing principle is to examine the pain of this regressing compared to the cost of running these tests (a few seconds per module currently translates to a few machine hours per month).

Another approach for reducing the cost, particularly for regression-searching (vs. correctness enforcing) tests, is to move tests into the periodic suite. However that entails failing changes promoting into development before later catching the issue, hopefully before they promote toward release.

Reply:
"I'm interested in the threshold where the detected cases stop being reliably detected (on say 1/10 runs) on your local environment. If that currently occurs around 50k iterations, then this is a reasonable test."
Kind of a tricky question to answer. If I run the unit tests without the shape/bus changes, I start to get < 100% reliability on Reference Component at 90 iterations, Capsule at 1000 iterations. Once I added multi-threading into the shape bus, these numbers went up to closer to the 10,000 - 15,000 range. Based solely on those numbers on my PC, we could _probably_ reduce all the iterations down to 10,000. However, if we tune the number of iterations to a specific CPU and a specific set of conditions that are now fixed, there's no particular reason to believe it is well-tuned to catch other sets of conditions that produce this result, or on other CPUs. To your point about "pain of regression", to me a few machine hours a month are an acceptable cost vs the cost of finding and fixing a regression if these components stop being threadsafe in some specific circumstance.

"Another approach for reducing the cost, particularly for regression-searching (vs. correctness enforcing) tests, is to move tests into the periodic suite."
Absolutely a fair suggestion. I'd prefer to keep these tests in the main suite due to the value I've already gotten from them though. Along with proving out the initial race condition, these unit tests also caught mistakes I made during implementation of this PR, including deadlocks that I accidentally introduced at one point due to some unexpected nested locks that could occur. Because they caught those problems, I feel like they fall more on the side of "correctness enforcing" for thread safety rather than just pure regression checking.

Reply:
My suggestion above is to use double that minimal threshold of where the local repro starts to become reliable, as a heuristic of where "good enough" lies. I'm comfortable with these tests living in the Main suite, but prefer that decision stay the hands of test-authors :)

Reply:
Sounds like a great heuristic! :) Reduced iterations on these tests to 30,000, which is 2X the 15,000 value that was the maximum needed to reliably reproduce the errors.